Introduce WTF::RefCountable

adetaylor opened the pull request:
#### 5edb8bb8be59ebb9f6b0447aab62c633fa0b22b9
<pre>
Introduce WTF::RefCountable
<a href="https://bugs.webkit.org/show_bug.cgi?id=303175">https://bugs.webkit.org/show_bug.cgi?id=303175</a>
<a href="https://rdar.apple.com/165481024">rdar://165481024</a>

Reviewed by Geoffrey Garen.

This introduces a new API, WTF::RefCountable, the purpose of which is to attach
a reference count to any type, potentially even move-only types.

This is needed for Swift/C++ interoperability because <a href="https://rdar.apple.com/162361370">rdar://162361370</a>
prevents us from passing move-only types from C++ to Swift. We&apos;ll need
to do that in the near future for WTF::Function and
WTF::CompletionHandler as we start to handle IPC message reception in
Swift code.

Fortunately, we already had such a type - as an implementation detail
of WTF::Box. Extract it from there to make it a standalone type.
Because this is the existing type, it retains the thread-safe
reference counting semantics.

Alternatives considered:
* Having a ref counted type for each thing we want to pass from C++
  to Swift, e.g. WTF::RefCounted::Function and
  WTF::RefCounted::CompletionHandler. Having a single RefCountable wrapper
  is obviously simpler.
* Naming this type SwiftCell, or similar. This has the advantage that
  it would be more obvious to remove such usages when
  <a href="https://rdar.apple.com/162361370">rdar://162361370</a> is fixed, but it was determined that the
  semantics of this type are not Swift-specific, so it should be
  named something general.
* Naming the type &apos;Cell&apos; or &apos;Ref::Cell&apos; or somesuch. This has been
  proposed in various discussions, but the code review process has
  favored &apos;RefCountable&apos; so that&apos;s where we&apos;ve ended up.

Canonical link: <a href="https://commits.webkit.org/303681@main">https://commits.webkit.org/303681@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4fc2a97d42b42de18613c70ad889249b7dadc8b4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133042 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5543 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44140 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140578 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85077 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/348ce136-918d-4dc6-8c85-5bce441d3756) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6015 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5406 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101743 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69088 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/52d6b4d1-939b-4e05-b7f5-e37f9ba122c0) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135988 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119221 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82542 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b7d906a5-d0f0-42fb-bae3-91d725042af5) 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1710 "Found 1 new API test failure: TestWebKitAPI.ObscuredContentInsets.TopOverhangColorExtensionLayer (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83813 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/125112 "Built successfully and passed tests") | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37335 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143231 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/131550 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5212 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37916 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110121 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5294 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4470 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110303 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28007 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4010 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115479 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58808 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5266 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33830 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/164518 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5107 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68718 "Built successfully") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/42881 "Build was cancelled. Recent messages:Failed to print configuration") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5356 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5224 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WTF/wtf/RefCountable.h:
```diff
@@ -0,0 +1,85 @@
+        return &m_value;
+    }
+
+#ifdef __swift__
+    void swiftRef()
+    {
+        WTF::ref(this);
+    }
+
+    void swiftDeref()
+    {
+        WTF::deref(this);
+    }
+#endif
```
In other refcounted types we make these free functions and remove the `#ifdef` and the word `swift`. e.g:

```
} SWIFT_SHARED_REFERENCE(refDataStore, derefDataStore);

}

inline void refDataStore(WebKit::WebsiteDataStore* WTF_NONNULL obj)
{
    WTF::ref(obj);
}

inline void derefDataStore(WebKit::WebsiteDataStore* WTF_NONNULL obj)
{
    WTF::deref(obj);
}
```

I think we should do that here, unless there's some reason we can't. While it's true that these functions are only needed because of Swift, we have a lot of functions that are only needed because of X, but I think the code is more understandable if the function just says what it does.

adetaylor replied:
OK! I'll adjust the names that in a follow-up PR because I already pressed merge here after your approval earlier this morning (and I'm kind of desperate to get this merged, as I have other PRs awaiting it).

*BUT* they can't be free functions, as that doesn't work in a templated type. They have to be member functions.

Annoyingly, we can't just do `SWIFT_SHARED_REFERENCE(.ref, .deref)` - we need to refer to member functions at this altitude in the class hierarchy.

I'll try a little harder to just call them `.ref` and `.deref` by having methods at this level which call the base class methods.

adetaylor committed:
Introduce WTF::RefCountable
https://bugs.webkit.org/show_bug.cgi?id=303175
rdar://165481024

Reviewed by Geoffrey Garen.

This introduces a new API, WTF::RefCountable, the purpose of which is to attach
a reference count to any type, potentially even move-only types.

This is needed for Swift/C++ interoperability because rdar://162361370
prevents us from passing move-only types from C++ to Swift. We'll need
to do that in the near future for WTF::Function and
WTF::CompletionHandler as we start to handle IPC message reception in
Swift code.

Fortunately, we already had such a type - as an implementation detail
of WTF::Box. Extract it from there to make it a standalone type.
Because this is the existing type, it retains the thread-safe
reference counting semantics.

Alternatives considered:
* Having a ref counted type for each thing we want to pass from C++
  to Swift, e.g. WTF::RefCounted::Function and
  WTF::RefCounted::CompletionHandler. Having a single RefCountable wrapper
  is obviously simpler.
* Naming this type SwiftCell, or similar. This has the advantage that
  it would be more obvious to remove such usages when
  rdar://162361370 is fixed, but it was determined that the
  semantics of this type are not Swift-specific, so it should be
  named something general.
* Naming the type 'Cell' or 'Ref::Cell' or somesuch. This has been
  proposed in various discussions, but the code review process has
  favored 'RefCountable' so that's where we've ended up.

Canonical link: https://commits.webkit.org/303681@main
